feat: added ISO18013 proximity flow (#187)
italia/eidas-it-wallet-docs: docs/en/relying-party-solution.rst

    The Relying Party specific metadata parameter are experimental 
    and still under discussion `here <https://github.com/openid/OpenID4VP/issues/17>`_.

Proximity Flow
==============

> This section describes how a Relying Party requests the presentation of an *mDoc-CBOR* Credential to a Wallet Instance according to the *ISO 18013-5 Specification*. Only *Supervised Device Retrieval flow* is supported in this technical implementation profile. 

Reviewer: ```suggestion
This section describes how a Verifier requests the presentation of an *mDoc-CBOR* Credential to a Wallet Instance according to the *ISO 18013-5 Specification*. Only *Supervised Device Retrieval flow* is supported in this technical implementation profile. 
```

let's use the term Verifier to distinguish an RP in proximity flow

The presentation phase is divided into three sub-phases: 
 

>   1. **Device Engagement**: This subphase starts when the User is asked to disclose some attributes from their mDoc(s). The objective of this subphase is to establish a secure communication channel between the Wallet Instance and the Verifier App, so that the mDoc requests and responses can be exchanged during the communication subphase.

Reviewer: ```suggestion
  1. **Device Engagement**: This subphase begins when the User is prompted to disclose certain attributes from the mDoc(s). The objective of this subphase is to establish a secure communication channel between the Wallet Instance and the Verifier App, so that the mDoc requests and responses can be exchanged during the communication subphase.
```

  The messages exchanged in this subphase are transmitted through short-range technologies to limit the possibility of interception and eavesdropping.

>   This technical implementation profile only support QR code for the Device Engagement.

Reviewer: ```suggestion
  This technical implementation profile exclusively supports QR code for Device Engagement.
```

>   2. **Session establishment**: In session establishment, the Verifier App establishes a connection on a secure channels. 
>   The data transmitted in such a connection will all be encrypted with the session key that both the Wallet Instance and the verifier know at this point in time. 

Reviewer: ```suggestion
  2. **Session establishment**: During the session establishment phase, the Verifier App sets up a secure connection. All data transmitted over this connection is encrypted using a session key, which is known to both the Wallet Instance and the Verifier at this stage.
```

  The established session MAY be terminated based on the conditions as detailed in [ISO18013-5#9.1.1.4].

>   3. **Communication - Device Retrieval**: The Verifier App encrypts the mdoc request with the appropriate session key and sends it to the Wallet Instance together with its public key in a session establishment message. The mdoc uses the data from the session establishment message to derive the session key and decrypt the mdoc request.

Reviewer: ```suggestion
  3. **Communication - Device Retrieval**: The Verifier App encrypts the mDoc request with the appropriate session key and sends it to the Wallet Instance together with its public key in a session establishment message. The mDoc uses the data from the session establishment message to derive the session key and decrypt the mDoc request.
```

>   In communication subphase, the Verifier App MAY ask the Wallet for information about the holder using mDoc requests and responses. 
>   The main channel for communication is the secure channel that has been established with the session establishment. The Wallet Instance encrypts the mdoc response with the appropriate session key and sends it to the Verifier App in a session data message.  
>   For the communication sub-phase, only BLE is supported by this technical implementation profile. 

Reviewer: ```suggestion
  During the communication subphase, the Verifier App has the option to request information from the Wallet using mDoc requests and responses. The primary mode of communication is the secure channel established during the session setup. The Wallet Instance encrypts the mDoc response using the session key and transmits it to the Verifier App via a session data message. This technical implementation profile only supports Bluetooth Low Energy (BLE) for the communication sub-phase.
```

> The Figure below presents the flow diagram of ISO 18013-5 compliant proximity flow.

Reviewer: ```suggestion
The following figure illustrates the flow diagram compliant with ISO 18013-5 for proximity flow.
```

.. _fig_High-Level-Flow-ITWallet-Presentation-ISO:
.. figure:: ../../images/High-Level-Flow-ITWallet-Presentation-ISO.svg
    :figwidth: 100%
    :align: center
    :target: https://www.plantuml.com/plantuml/svg/bL9BZnCn3BxFhx3A0H3q3_ImMlOXXBJYqGguzE9ct2RQn0bvJDb_ZoSP3QFI2xab_Xx-xDocZ34NPpiisNDn1ufT1t9GPH_XUw88cA3KjuF_3QlnwNM2dHDYq9vf1Q-Up4ddErkeme9KZ381ESFg9rfB6JwnEB4IiAYTAuou7nN_Al-WQ8xcVzHd2dm8eKeFI-cMfApNDpVd3Nm9n90rmKLBa3s4I8b441dSWrTm7wcNkq7RD3xxJE07CIhlXmqyq624-CWdF94RYQaSWiP4iAweRzjr1vLvRkOVYIcYY32TWO8c9rSBp_GYWKoSe88LzPtsvx5HKO5xtnCSVVpNibA6ATjE8IyfKr7aBgptVDry0WlPXIBOH2aPpoEcbgzDOJTXIEPui2PfrqROZogki56OfNuvcxkdHv5N9H8eZSnaPLRJwUPU95JTn9P-5J60Tn2AcAZQjJ_MiCljxndUN6texN8Dr-ErSjd0roZrNEUjFDSVaJqaZP6gOMpDK0-61UHglkcJjJL75Cx4NHflAKT30xLGH_41wnLQIDb7FD6C7URSAOZCSfCjxyjSWcHEZBb4slCuTQL9FJVsWDRq9akuxfQuByx-0G00

    High-Level Proximity Flow

> **Step 1-3**: The Verifier asks the User to disclose some attributes from their mdoc(s) stored in the Wallet Instance. The User opens the Wallet Instance. The Wallet Instance MUST generate a new ephemeral key pair (EDeviceKey.Priv, EDeviceKey.Pub), and includes the cipher suite identifier, the identifier of the elliptic curve to be used for key agreement and the EDeviceKey public point, as part of the device engagement structure (see [ISO18013-5#9.1.1.4]). This pair of keys is ephemeral and MUST be invalidated as soon as the secure channel is established. Finally, the Walle Instance shows the QR Code for the Device Engagement.

Reviewer: ```suggestion
**Step 1-3**: The Verifier requests the User to reveal certain attributes from their mDoc(s) stored in the Wallet Instance. The User initiates the Wallet Instance. The Wallet Instance MUST create a new temporary key pair (EDeviceKey.Priv, EDeviceKey.Pub), and incorporate the cipher suite identifier, the identifier of the elliptic curve for key agreement, and the EDeviceKey public point into the device engagement structure (refer to [ISO18013-5#9.1.1.4]). This key pair is temporary and MUST be invalidated immediately after the secure channel is established. Finally, the Wallet Instance displays the QR Code for Device Engagement.
```

> Following is an example of a device engagement structure for QR device engagement and BLE as the data retrieval mode.

Reviewer: ```suggestion
Below an example of a device engagement structure that utilizes QR for device engagement and Bluetooth Low Energy (BLE) for data retrieval.
```

  a30063312e30018201d818584ba4010220012158205a88d182bce5f42efa59943f33359d2e8a968ff289d93e5fa444b624343167fe225820b16e8cf858ddc7690407ba61d4c338237a8cfcf3de6aa672fc60a557aa32fc670281830201a300f401f50b5045efef742b2c4837a9a3b0e1d05a6917 

In diagnostic notation:

.. code-block:: 

  { 
    0: "1.0", % Version

    1:        % Security
    [ 
        1,     % defines the cipher suite 1 which contains only EC curves
        24(<<  % embedded CBOR data item
          { 
            1: 2, % kty:EC2 (Elliptic curves with x and y coordinate pairs)
          -1: 1, % crv:p256
  -2:h'5A88D182BCE5F42EFA59943F33359D2E8A968FF289D93E5FA444B624343  167FE',% x-coordinate
  -3:h'B16E8CF858DDC7690407BA61D4C338237A8CFCF3DE6AA672FC60A557AA32FC67' % y-coordinate
          }
        >>)
      ],
  
      2: %DeviceRetrievalMethods(Device engagement using QR code)
      [ 
        [
          2, %BLE
          1, % Version
        {    %BLE options
            0: false, % no support for mdoc peripheral server mode
            1: true, % support mdoc central client mode
            11: h'45EFEF742B2C4837A9A3B0E1D05A6917' % UUID of mdoc client central mode
          }
        ]
      ]
  }

> **Step 4-6**: The Verifier App scan the QR Code, generates its own set of ephemeral keys (EReaderKey. Priv, EReaderKey.Pub) and computes the session key, involving the public key received in the Engagement Structure and its newly-generated private key as specified in [ISO18013-5#9.1.1.5]. Finally, generates its session key that MUST be derived independently by the Wallet Instance and the Verifier App.

Reviewer: ```suggestion
**Step 4-6**: The Verifier App scans the QR Code and generates its own ephemeral key pair (EReaderKey.Priv, EReaderKey.Pub). It then calculates the session key, using the public key received in the Engagement Structure and its newly-generated private key, as outlined in [ISO18013-5#9.1.1.5]. Finally, it generates its session key, which must be independently derived by both the Wallet Instance and the Verifier App.
```

> **Step 7**: The Verifier App generates the mdoc request that MUST be encrypted with the appropriate session key, and sends it to the Wallet Instance together with EReaderKey.Pub in a session establishment message. The mdoc request MUST be encoded as CBOR as in the following  non-normative example.

Reviewer: ```suggestion
**Step 7**: The Verifier App creates an mDoc request that MUST be encrypted using the relevant session key, and transmits it to the Wallet Instance along with EReaderKey.Pub within a session establishment message. The mDoc request MUST be encoded in CBOR, as demonstrated in the following non-normative example.
```

> The above CBOR data in diagnostic notation: 

Reviewer: ```suggestion
The above CBOR data is represented in diagnostic notation as follows:
```

            "issue_date": false,
            "issuing_country": false,
            "issuing_authority": false,
            "birth_date": false,
            "expiry_date": false,
            "given_name": false,
            "portrait": false,
            "family_name": false
          }
        }
      } >>),
      "readerAuth": [
        h'a10126',
        {
          33: h'308201253081cda00302010202012a300a06082a8648ce3d0403023020311e301c06035504030c15536f6d652052656164657220417574686f72697479301e170d3233313132343130323832325a170d3238313132323130323832325a301a3118301606035504030c0f536f6d6520526561646572204b65793059301306072a8648ce3d020106082a8648ce3d03010703420004aa1092fb59e26ddd182cfdbc85f1aa8217a4f0fae6a6a5536b57c5ef7be2fb6d0dfd319839e6c24d087cd26499ec4f87c8c766200ba4c6218c74de50cd1243b1300a06082a8648ce3d0403020347003044022048466e92226e042add073b8cdc43df5a19401e1d95ab226e142947e435af9db30220043af7a8e7d31646a424e02ea0c853ec9c293791f930bf589bee557370a4c97b'
        },
        null,
        h'58a0d421a7e53b7db0412a196fea50ca6d4c8a530a47dd84d88588ab145374bd0ab2a724cf2ed2facf32c7184591c5969efd53f5aba63194105440bc1904e1b9'
      ]
    }
    ]
  }

> **Step 8**: The Wallet Instance uses the session establishment message to derive the session keys and decrypt the mdoc request. It computes the session key using the public key received from the Verifier App and its private key.

Reviewer: ```suggestion
**Step 8**: The Wallet Instance uses the session establishment message to derive the session keys and decrypt the mDoc request. It computes the session key using the public key received from the Verifier App and its private key.
```

> **Step 9-10**: Upon receiving the mDoc request, the Wallet Instance identifies the documents containing the requested attributes and prompts the User for permission to share this information with the Verifier. If the User consents, the Wallet produces an mDoc response and sends it to the Verifier App via the secure channel.

Reviewer: ```suggestion
**Step 9-10**: When the Wallet Instance receives the mDoc request, it locates the documents that contain the requested attributes and asks the User for permission to provide this information to the Verifier. If the User agrees, the Wallet generates an mDoc response and transmits it to the Verifier App through the secure channel.
```

> **Step 11-12**: In case of User consent, the Wallet Instance generates an mDoc response and sends it to the Verifier App through the secure channel. The mDoc response MUST be CBOR encoded, and its structure is defined in [ISO18013-5#8.3.2.1.2.2]. A non-normative example of a mDoc response is given below.

Reviewer: ```suggestion
**Step 11-12**: If the User gives consent, the Wallet Instance creates an mDoc response and transmits it to the Verifier App via the secure channel. The mDoc response MUST be encoded in CBOR, with its structure outlined in [ISO18013-5#8.3.2.1.2.2]. Below is a non-normative example of an mDoc response.
```

          4: h'4A4B4CC64EC9299C1A2501EA449F577005E9F7A60408057C07A7C67FB151E5F5',
          11: h'78824FBD6FBBA88A2AAB44DF8B6F5E9759126D87D1F4415995E658FD9239E1FE'
          }
        },
        "deviceKeyInfo": {
          "deviceKey": {
          1: 2,
          -1: 1,
          -2: h'AFD09E720B918CEDC2B8A881950BAB6A1051E18AE16A814D51E609938663D5E1',
          -3: h'61FBC6C8AD24EC86A78BB4E9AC377DD2B7C711D9F2EB9AFD4AA0963662847AED'}},
          "validityInfo": {
            "signed": 0("2023-11-24T14:54:05Z"),
            "validFrom": 0("2023-11-24T14:54:05Z"),
            "validUntil": 0("2024-11-24T14:54:05Z")}
          }  >>),
          h'f2461e4fab69e9f7bcffe552395424514524d1679440036213173101448d1b1ab4a293859b389ffa8b47aeed10e9b0c1545412ac37c51a76482cd9bbbe110152'
        ]
      },
      "deviceSigned": {
        "nameSpaces": 24(<< {} >>),
        "deviceAuth": {
          "deviceSignature": [
            h'a10126',
            {},
            null,
            h'1fed7190d2975ab79c072e6f1d9d52436059d1fc959d55baf74f057d89b10fcc0dc77a50d433d4c76ddf26223c5560c4ab123b5cb5eb805a90036aa147493076'
          ]
        }
      }
    }
    ],
    "status": 0
  }

> **Step 13**: The Verifier App MUST verify that the signatures in the mDoc **issuerSigned** field are valid using the public key of the Credential Issuer declared in the mDoc itself.  At this point, the verifier MUST check the signature in the **deviceSigned** field. If the check of these signatures is successful, the verifier can trust that the information it received is valid.

Reviewer: ```suggestion
**Step 13**: The Verifier App is required to validate the signatures in the mDoc's issuerSigned field using the public key of the Credential Issuer specified within the mDoc. Subsequently, the Verifier MUST validate the signature in the deviceSigned field. If these signature checks pass, the Verifier can confidently consider the received information as valid.
```

> 
> 
> 
> 

Reviewer: ```suggestion
```

Device Engagement
-----------------

The Device Engagement structure MUST be have at least the following components:

  - **Version**: *tstr*. Version of the data structure being used.
  - **Security**: an array that contains two mandatory values
  
    - the cipher identifier: see Table 22 of [ISO18013-5]
    - the mDL public ephemeral key generated by the Wallet Instance and required by the Verifier App to derive the Session Key. The mDL public ephemeral key MUST be of a type allowed by the indicated cipher suite.

>   - **TransferMethod**: an array that contains one or more transferMethod arrays when performing device engagement using the QR code. This array is for offline data retrieval methods. A transferMethod array holds two mandatory values (type and version). Only the BLE option is supported by this technical implementation profile, then the type value MUST be set to '2'. 

Reviewer: ```suggestion
  - **transferMethod**: an array that contains one or more transferMethod arrays when performing device engagement using the QR code. This array is for offline data retrieval methods. A transferMethod array holds two mandatory values (type and version). Only the BLE option is supported by this technical implementation profile, then the type value MUST be set to ``2``. 
```

            - **elementIdentifier**: (tstr). Identifier of User attribute name contained in the Credential.
            - **elementValue**: (any). User attribute value
      - **DeviceSigned**: Responded data elements signed by the Wallet Instance.

        - **NameSpaces**: #6.24(bstr .cbor DeviceNameSpaces). The DeviceNameSpaces structure MAY be an empty structure. DeviceNameSpaces contains the data element identifiers and values. It is returned as part of the corresponding namespace in DeviceNameSpace.

          - **DataItemName**: (tstr). The identifier of the element.
          - **DataItemValue**: (any). The value of the element.
        - **DeviceAuth**:  The DeviceAuth structure MUST contain the DeviceSignature elements.

          - **DeviceSignature**: It MUST contain the device signature for the Wallet Instance authentication. 
  - **status**: It contains a status code. For detailed description and action required refer to to Table 8 (ResponseStatus) of the [ISO18013-5]

Session Termination
-------------------

The session MUST be terminated if at least one of the following conditions occur. 

>   - After a time-out of no activity of receiving or sending session establishment or session data messages occurs. The time-out for no activity implemented by the Wallet Instance and the Verifier App SHOULD be no less than 300 s.

Reviewer: ```suggestion
  - After a time-out of no activity of receiving or sending session establishment or session data messages occurs. The time-out for no activity implemented by the Wallet Instance and the Verifier App SHOULD be no less than 300 seconds.
```

>   - If the Wallet Instance does not want to receive any further requests. 

Reviewer: ```suggestion
  - When the Wallet Instance doesn't accept any more requests.
```

>   - If the Wallet Instance does not want to receive any further requests. 

Reviewer: ```suggestion
  - When the Wallet Instance does not receive any further requests. 
```

>   - If the Verifier App does not want to send any further requests. 

Reviewer: ```suggestion
  - When the Verifier App does not send any further requests. 
```

> If the Wallet Instance and the Verifier App does not want to send or receive any further requests, it MUST initiate session termination as follows. 

Reviewer: ```suggestion
If the Wallet Instance and the Verifier App does not send or receive any further requests, the session termination MUST be initiated as follows. 
```

>   - to send the status code for session termination or 
>   - to send the "End" command defined in [ISO18013-5#8.3.3.1.1.5].

Reviewer: ```suggestion
 - Send the status code for session termination, or
 - dispatch the "End" command as outlined in [ISO18013-5#8.3.3.1.1.5].
```
